### ISSUE TITLE
`this` scoping issue

### ISSUE THREAD

**Issue by freeman42x** [vickychijwani/closer.js#16]
It seems that in CodeCombat the functions operating on `this` usually work when called from outside functions or inside functions, and usually they do not work when nested deeper into code.

For example:

```
(dotimes [n 1000]
  (let [enemy (.findNearestEnemy this)]
    (.say this "Is there an enemy?")
    (if enemy
      (do
        (.say this "An enemy spotted!")
        (if (< (.distanceTo this enemy) 5)
          (.attack this enemy)
          (.shield this)))
      (.moveXY this 40 34))))
```

will print `Is there an enemy?` but never `An enemy spotted!` even though it becomes true.

This might have something to do with the visibility of `this` / JavaScript scoping rules.

Since I can only log using `.say` and `.say` needs `this` I can't even log `this` to figure out if `this` happens to be nil there.

**Comment by razvan-panda** [comment]
If I use:

(dotimes [n 1000]
      (let [enemy (.findNearestEnemy this)]
        (.say this "Is there an enemy?")
        (if enemy
          (.say this "An enemy spotted!")
          (.moveXY this 40 34))))

it will display correctly "An enemy spotted!" when condition becomes true.

So it seems like the `do` is doing something to `this`, pardon the bad pwn.

**Comment by vickychijwani** [comment]
Yes, this is an issue I'm aware of, and yes, you are right that it is caused by JavaScript's rules around `this`. `do` creates a new JavaScript function scope, which is why the value of `this` changes.

**Comment by razvan-panda** [comment]
I used a function definition to get past the scoping issue:

    (defn attack [enemy]
      (if (< (.distanceTo this enemy) 5)
        (.attack this enemy)
        (.shield this)))
    
    (dotimes [n 1000]
      (let [enemy (.findNearestEnemy this)]
        (if enemy
          (attack enemy)
          (.moveXY this 40 34))))

Are there any temporary hacks to get `this` usable inside the `do`?

**Comment by razvan-panda** [comment]
Same issue in `if`s:

    (dotimes [n 1000]
      (def enemy (.findNearestEnemy this))
      (if enemy
        (if (< (.distanceTo this enemy) 5)
          (.say this "message")
          (if (.isReady this "cleave")
            (.cleave this enemy)
            (.attack this enemy)))))

"message" never gets said.

**Comment by razvan-panda** [comment]
Another scoping oddity, this works:

    (let [x this]
      (dotimes [n 1000]
        (def enemy (.findNearestEnemy x))
        (.say x (str enemy))
        (if enemy
          (if (< (.distanceTo x enemy) 5)
            (if (.isReady x "cleave")
              (.cleave x enemy)
              (.attack x enemy))))))

if ` (.say x (str enemy))` is removed then the attacks no longer work.
